internal/cssc/cssc.go

```diff
@@ -124,37 +205,70 @@ func ApplyFilterAndGetFilteredList(ctx context.Context, acrClient api.AcrCLIClie
 		if err != nil {
 			var listTagsErr *tag.ListTagsError
 			if errors.As(err, &listTagsErr) {
+				for _, tag := range filterRepo.Tags {
+					artifactsNotFound = append(artifactsNotFound, FilteredRepository{
+						Repository: filterRepo.Repository,
+						Tag:        tag,
+					})
+				}
 				continue
 			}
-			return nil, errors.Wrap(err, "Some unexpected error occurred while listing tags for repository-"+filterRepo.Repository)
+			return nil, nil, errors.Wrap(err, "Some unexpected error occurred while listing tags for repository-"+filterRepo.Repository)
 		}
 
 		if len(filterRepo.Tags) == 1 && filterRepo.Tags[0] == "*" { // If the repo has * as tags defined in the filter, then all tags are considered for that repo
 			for _, tag := range tagList {
-				if strings.HasSuffix(*tag.Name, "-patched") {
-					originalTag := (*tag.Name)[:len(*tag.Name)-len("-patched")]
-					matchingRepo := FilteredRepository{Repository: filterRepo.Repository, Tag: originalTag, PatchTag: *tag.Name}
-					filteredRepos = AppendElement(filteredRepos, matchingRepo)
-				} else {
-					matchingRepo := FilteredRepository{Repository: filterRepo.Repository, Tag: *tag.Name, PatchTag: *tag.Name}
-					filteredRepos = AppendElement(filteredRepos, matchingRepo)
+				matches := patchTagRegex.FindStringSubmatch(*tag.Name)
+				if matches != nil {
+					baseTag := matches[1]
+					tagMap[baseTag] = append(tagMap[baseTag], *tag.Name)
+				} else if !endsWithIncrementalOrFloatingPattern(*tag.Name) {
+					tagMap[*tag.Name] = append(tagMap[*tag.Name], *tag.Name)
 				}
 			}
 		} else {
 			for _, ftag := range filterRepo.Tags { // If the repo has specific tags defined in the filter, then only those tags are considered
+				// If tag from filter is not found in the tag list obtained for the repository, then add it to artifactsNotFound and continue with the next tag
+				if !containsTag(tagList, ftag) {
+					artifactsNotFound = append(artifactsNotFound, FilteredRepository{
+						Repository: filterRepo.Repository,
+						Tag:        ftag,
+					})
+					continue
+				}
 				for _, tag := range tagList {
-					if *tag.Name == ftag {
-						matchingRepo := FilteredRepository{Repository: filterRepo.Repository, Tag: *tag.Name, PatchTag: *tag.Name}
-						filteredRepos = AppendElement(filteredRepos, matchingRepo)
-					} else if *tag.Name == ftag+"-patched" {
-						matchingRepo := FilteredRepository{Repository: filterRepo.Repository, Tag: ftag, PatchTag: ftag + "-patched"}
-						filteredRepos = AppendElement(filteredRepos, matchingRepo)
+					// This is needed to evaluate all versions of patch tags when original tag is specified in the filter
+					re := regexp.MustCompile(`^` + ftag + `(-patched\d*)?$`)
```

Comment:
If a customer uses similar tags across repos you will end up recompiling the regex a bunch of times. Also note that mustCompile is not typically used with dynamic regexes as it panics on error. Id suggest caching it.

Reply:
Moved regex compilation outside the inner loop to compile regex once per ftag and reuse it inside the tagList loop